File: chat/tools/feedback_intent.py
"""
Rule-based feedback intent classifier + reply window (mirrors Node feedbackIntent.js).
Used when Python session is in awaiting_feedback_* steps.
"""

from __future__ import annotations

import os
import re
from datetime import datetime, timezone
from typing import Any

FEEDBACK_REPLY_WINDOW_MINUTES = max(
    1, int(os.getenv("FEEDBACK_REPLY_WINDOW_MINUTES", "15"))
)
FEEDBACK_REPLY_WINDOW_SECONDS = FEEDBACK_REPLY_WINDOW_MINUTES * 60

_RESET = frozenset({
    "home", "menu", "main menu", "mainmenu", "restart", "start over", "startover",
    "reboot", "new", "begin",
})
_SERVICE_MENU_IDS = frozenset({
    "dine_in", "takeaway_now", "takeaway_schedule", "takeaway", "delivery",
    "schedule_delivery", "reserve_table", "book_table", "order_food",
})
_RATING_MAP = {
    "excellent": 5, "5": 5, "good": 4, "4": 4, "average": 3, "3": 3,
    "below average": 2, "below_average": 2, "2": 2, "poor": 1, "1": 1,
}
_SKIP = frozenset({
    "skip", "s", "none", "no", "done", "ok", "okay", "skip_aspects", "skip_comment",
})
_GREETING_RE = re.compile(
    r"^(hi|hello|hey|hola|namaste|good\s+(morning|afternoon|evening)|gm|yo)\b",
    re.IGNORECASE,
)
_ORDER_HINT_RE = re.compile(
    r"\b(menu|order|idli|dosa|biryani|cart|done|catalog|takeaway|delivery|dine)\b",
    re.IGNORECASE,
)

# Map Python booking_step → classifier sub-state
_STEP_TO_SUBSTATE = {
    "awaiting_feedback_rating": "awaiting_feedback_rating",
    "awaiting_feedback_aspects": "awaiting_feedback_tags",
    "awaiting_feedback_comment": "awaiting_feedback_comment",
}


def _normalize(text: str) -> str:
    t = (text or "").strip().lower()
    if t.startswith("⏭️"):
        t = t[1:].strip()
    return t


def _parse_rating(text: str) -> int | None:
    raw = _normalize(text)
    if raw in _RATING_MAP:
        return int(_RATING_MAP[raw])
    if re.fullmatch(r"[1-5]", raw):
        return int(raw)
    m = re.search(r"\b([1-5])\b", raw)
    if m:
        return int(m.group(1))
    stars = text.count("⭐") + text.count("★")
    if 1 <= stars <= 5 and len(text.strip()) <= 8:
        return stars
    return None


def _is_tag_selection(text: str) -> bool:
    t = _normalize(text)
    if t in {"all", "everything", "all of the above"}:
        return True
    return bool(re.fullmatch(r"[\d\s,;]+", t) and re.search(r"\d", t))


def _extract_interactive_id(message_obj: dict[str, Any] | None) -> str:
    if not message_obj or not isinstance(message_obj, dict):
        return ""
    if message_obj.get("type") != "interactive":
        return ""
    interactive = message_obj.get("interactive") or {}
    if interactive.get("type") == "list_reply":
        lr = interactive.get("list_reply") or {}
        return (lr.get("id") or lr.get("title") or "").strip()
    if interactive.get("type") == "button_reply":
        br = interactive.get("button_reply") or {}
        return (br.get("id") or br.get("title") or "").strip()
    return ""


def classify_session_feedback_intent(
    text: str,
    booking_step: str,
    message_obj: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Returns intent, feedback_action, abandon_feedback_flow."""
    sub_state = _STEP_TO_SUBSTATE.get(booking_step, booking_step)
    raw = (text or "").strip()
    lower = _normalize(raw)

    def abandon(intent: str) -> dict[str, Any]:
        return {
            "intent": intent,
            "feedback_action": None,
            "abandon_feedback_flow": True,
        }

    def reply(action: str) -> dict[str, Any]:
        return {
            "intent": "feedback_reply",
            "feedback_action": action,
            "abandon_feedback_flow": False,
        }

    if message_obj and message_obj.get("type") == "order":
        return abandon("new_order")

    iid = _extract_interactive_id(message_obj)
    if iid:
        il = iid.lower()
        if il in _SERVICE_MENU_IDS:
            return abandon("menu_action")
        if il in _SKIP:
            return reply("skip")
        if sub_state == "awaiting_feedback_rating" and il in _RATING_MAP:
            return reply("rating")
        if sub_state == "awaiting_feedback_rating":
            return abandon("menu_action")
        return abandon("menu_action")

    if lower in _RESET or lower in _SERVICE_MENU_IDS:
        return abandon("menu_action")
    if _GREETING_RE.match(lower):
        return abandon("greeting")
    if lower in _SKIP:
        return reply("skip")

    if sub_state == "awaiting_feedback_rating":
        if re.fullmatch(r"[\d\s,;]+", lower) and re.search(r"[\s,;]", lower) and re.search(r"\d", lower):
            return abandon("other")
        if _parse_rating(raw) is not None:
            return reply("rating")
        if _ORDER_HINT_RE.search(raw):
            return abandon("new_order")
        return abandon("other")

    if sub_state == "awaiting_feedback_tags":
        if _is_tag_selection(raw):
            return reply("tags")
        if _ORDER_HINT_RE.search(raw):
            return abandon("new_order")
        return abandon("other")

    if sub_state == "awaiting_feedback_comment":
        if raw:
            return reply("comment")
        return abandon("other")

    return abandon("other")


def is_session_feedback_expired(session_state: dict[str, Any]) -> bool:
    sent_at = session_state.get("feedback_invite_sent_at")
    if not sent_at:
        return False
    try:
        if isinstance(sent_at, (int, float)):
            start = datetime.fromtimestamp(sent_at, tz=timezone.utc)
        else:
            start = datetime.fromisoformat(str(sent_at).replace("Z", "+00:00"))
        elapsed = (datetime.now(timezone.utc) - start).total_seconds()
        return elapsed > FEEDBACK_REPLY_WINDOW_SECONDS
    except Exception:
        return False


def is_db_feedback_invite_active(feedback_sent_at: str | None) -> bool:
    """True if Node feedback_pending invite is still inside the reply window."""
    if not feedback_sent_at:
        return True
    try:
        start = datetime.fromisoformat(str(feedback_sent_at).replace("Z", "+00:00"))
        elapsed = (datetime.now(timezone.utc) - start).total_seconds()
        return elapsed <= FEEDBACK_REPLY_WINDOW_SECONDS
    except Exception:
        return True


def clear_session_feedback(session_state: dict[str, Any]) -> None:
    """Drop feedback sub-state so normal booking routing resumes."""
    for key in (
        "feedback_booking_id", "feedback_token", "feedback_table",
        "feedback_rating", "feedback_rating_label", "feedback_aspects",
        "feedback_comment", "_feedback_aspects_list", "_feedback_flow_token",
        "feedback_invite_sent_at",
    ):
        session_state.pop(key, None)
    session_state["booking_step"] = "visit_complete"

File: src/helpers/feedbackIntent.js
// src/helpers/feedbackIntent.js
// Reply-window gate + rule-based intent classification for post-visit feedback.
// No LLM calls — threshold check runs first; ambiguous in-window messages abandon
// the flow rather than blocking normal ordering.

'use strict';

const { supabaseAdmin } = require('../config/supabase');
const { phoneVariants } = require('./conversationState');

const FEEDBACK_REPLY_WINDOW_MS =
  Math.max(1, parseInt(process.env.FEEDBACK_REPLY_WINDOW_MINUTES || '15', 10)) *
  60 *
  1000;

const RESET_KEYWORDS = new Set([
  'home', 'menu', 'main menu', 'mainmenu', 'restart', 'start over', 'startover',
  'reboot', 'new', 'begin',
]);

const SERVICE_MENU_IDS = new Set([
  'dine_in', 'takeaway_now', 'takeaway_schedule', 'takeaway', 'delivery',
  'schedule_delivery', 'reserve_table', 'book_table', 'order_food',
]);

const GREETING_RE = /^(hi|hello|hey|hola|namaste|good\s+(morning|afternoon|evening)|gm|yo)\b/i;

const RATING_MAP = {
  excellent: 5, 5: 5,
  good: 4, 4: 4,
  average: 3, 3: 3,
  'below average': 2, below_average: 2, 2: 2,
  poor: 1, 1: 1,
};

const SKIP_TOKENS = new Set([
  'skip', 's', 'none', 'no', 'done', 'ok', 'okay', 'skip_aspects', 'skip_comment',
]);

const ORDER_HINT_RE =
  /\b(menu|order|idli|dosa|biryani|cart|done|catalog|takeaway|delivery|dine)\b/i;

function normalizeText(text) {
  return String(text || '').trim().toLowerCase().replace(/^⏭️\s*/, '');
}

function getReplyWindowStart(record) {
  return record.feedback_sent_at || record.freed_at || record.updated_at || null;
}

function isReplyWindowExpired(record) {
  const start = getReplyWindowStart(record);
  if (!start) return false;
  return Date.now() - new Date(start).getTime() > FEEDBACK_REPLY_WINDOW_MS;
}

/** Map feedback_pending row → sub-state name used by the classifier spec. */
function getFeedbackSubState(record, aspectsPayloadFn) {
  if (record.feedback_rating == null) return 'awaiting_feedback_rating';
  if (!aspectsPayloadFn(record)) return 'awaiting_feedback_tags';
  return 'awaiting_feedback_comment';
}

function parseRating(text) {
  const raw = normalizeText(text);
  if (RATING_MAP[raw] != null) return RATING_MAP[raw];
  if (/^\s*[1-5]\s*$/.test(raw)) return parseInt(raw, 10);
  const digit = raw.match(/\b([1-5])\b/);
  if (digit) return parseInt(digit[1], 10);
  const stars = String(text || '').match(/[⭐★]/g);
  if (stars?.length) return Math.min(stars.length, 5);
  return null;
}

function isSkipToken(text) {
  return SKIP_TOKENS.has(normalizeText(text));
}

function isTagSelection(text) {
  const t = normalizeText(text);
  if (['all', 'everything', 'all of the above'].includes(t)) return true;
  return /^[\d\s,;]+$/.test(t) && /\d/.test(t);
}

function extractInteractiveId(message) {
  if (!message || typeof message !== 'object') return '';
  if (message.type === 'interactive') {
    const interactive = message.interactive || {};
    if (interactive.type === 'list_reply') {
      return (interactive.list_reply?.id || interactive.list_reply?.title || '').trim();
    }
    if (interactive.type === 'button_reply') {
      return (interactive.button_reply?.id || interactive.button_reply?.title || '').trim();
    }
  }
  return '';
}

/**
 * Rule-based intent classifier (spec-aligned JSON).
 * @returns {{ intent: string, feedback_action: string|null, abandon_feedback_flow: boolean }}
 */
function classifyFeedbackIntent({ text, message, subState }) {
  const raw = String(text || '').trim();
  const lower = normalizeText(raw);
  const abandon = (intent, action = null) => ({
    intent,
    feedback_action: action,
    abandon_feedback_flow: true,
  });
  const reply = (action) => ({
    intent: 'feedback_reply',
    feedback_action: action,
    abandon_feedback_flow: false,
  });

  if (message?.type === 'order') {
    return abandon('new_order');
  }

  const interactiveId = extractInteractiveId(message);
  const interactiveLower = interactiveId.toLowerCase();

  if (interactiveId) {
    if (SERVICE_MENU_IDS.has(interactiveLower)) {
      return abandon('menu_action');
    }
    if (SKIP_TOKENS.has(interactiveLower)) {
      return reply('skip');
    }
    if (subState === 'awaiting_feedback_rating' && RATING_MAP[interactiveLower] != null) {
      return reply('rating');
    }
    if (subState === 'awaiting_feedback_rating' && !RATING_MAP[interactiveLower]) {
      return abandon('menu_action');
    }
    if (subState !== 'awaiting_feedback_rating') {
      return abandon('menu_action');
    }
  }

  if (RESET_KEYWORDS.has(lower) || SERVICE_MENU_IDS.has(lower)) {
    return abandon('menu_action');
  }
  if (GREETING_RE.test(lower)) {
    return abandon('greeting');
  }
  if (isSkipToken(raw)) {
    return reply('skip');
  }

  if (subState === 'awaiting_feedback_rating') {
    if (/^[\d\s,;]+$/.test(lower) && /[\s,;]/.test(lower) && /\d/.test(lower)) {
      return abandon('other');
    }
    if (parseRating(raw) != null) return reply('rating');
    if (ORDER_HINT_RE.test(raw)) return abandon('new_order');
    return abandon('other');
  }

  if (subState === 'awaiting_feedback_tags') {
    if (isTagSelection(raw)) return reply('tags');
    if (ORDER_HINT_RE.test(raw)) return abandon('new_order');
    return abandon('other');
  }

  if (subState === 'awaiting_feedback_comment') {
    if (raw.length > 0) return reply('comment');
    return abandon('other');
  }

  return abandon('other');
}

/**
 * Mark feedback invite closed without manager alert (window expired or abandoned).
 */
async function gracefullyExpireFeedback(record, reason = 'expired:window') {
  if (!record?.id) return;

  await supabaseAdmin
    .from('feedback_pending')
    .update({
      manager_notified: true,
      feedback_received_at: new Date().toISOString(),
      feedback_text: reason,
    })
    .eq('id', record.id);

  const variants = phoneVariants(record.customer_phone);
  if (variants.length) {
    await supabaseAdmin
      .from('feedback_pending')
      .update({
        manager_notified: true,
        feedback_received_at: new Date().toISOString(),
        feedback_text: reason,
      })
      .eq('restaurant_id', record.restaurant_id)
      .in('customer_phone', variants)
      .eq('feedback_sent', true)
      .eq('manager_notified', false);
  }

  const { syncConversationForFeedbackComplete } = require('./conversationState');
  await syncConversationForFeedbackComplete({
    restaurantId: record.restaurant_id,
    customerPhone: record.customer_phone,
  }).catch(() => {});
}

module.exports = {
  FEEDBACK_REPLY_WINDOW_MS,
  classifyFeedbackIntent,
  getFeedbackSubState,
  getReplyWindowStart,
  gracefullyExpireFeedback,
  isReplyWindowExpired,
  parseRating,
  RATING_MAP,
  SERVICE_MENU_IDS,
};
